### PR TITLE
Restructure repository

### PR DESCRIPTION
**Pull request by RussKie**
* Rename Setup/ to setup/
* Rename Bin/ into assets/ and move under setup/
* Rename Externals/ to externals/
* Move app projects into src/
* Move test projects into tests/
* Move scripts/ to eng/
* Move TranslationApp/ into setup/

### CODE REVIEW

File: native/GitExtensionsShellEx/Generated/GitExtensionsShellEx.h
Comment (gerhardol) [inline native/GitExtensionsShellEx/Generated/GitExtensionsShellEx.h:1]: Remove generated files?

Comment (gerhardol) [inline native/GitExtensionsShellEx/Generated/GitExtensionsShellEx.h:1]: Why separating native?
What "standard" is this based on?

**Review by gerhardol — COMMENTED** [review]
In general I am not fond of moving files around, it breaks patches and history.

**Comment by RussKie** [comment]
> In general I am not fond of moving files around, it breaks patches and history.

Can't disagree, though renames/moves should work more or less. 

The root of the repo is very messy, which is confusing, and it could be a barrier for prospective contributors. 

This has been bothering me for many years, and this is my Nth attempt, which I finally pushed in.

File: native/GitExtensionsShellEx/Generated/GitExtensionsShellEx.h
Comment (RussKie) [reply native/GitExtensionsShellEx/Generated/GitExtensionsShellEx.h:1]: None. These projects are C++, and are not part of the main solution. Having them in the same folder as the C# projects, which are part of the solution, feels confusing. These projects are built via native.proj too (I should move it from eng/ here). So, the name "native" stemmed from there.

Do you have better suggestions?

Comment (gerhardol) [reply native/GitExtensionsShellEx/Generated/GitExtensionsShellEx.h]: move below src/, as "top" level project feels natural to me
Alternative is to add yet another layer with app/, shellext/ below src/

**Comment by gerhardol** [comment]
I will approve if you rebase #10906 and consider taking it out of draft - it works fine as is (also mstv workaround)

**Comment by RussKie** [comment]
> I will approve if you rebase #10906 and consider taking it out of draft - it works fine as is (also mstv workaround)

Sure thing. 
I've rebased that branch on top of this one (there was only one file - FormBrowse.Designer.cs - that I had to manually re-apply) but I'm not going to push it in until this branch is merged to avoid making a mess of that PR.

File: native/GitExtensionsShellEx/Generated/GitExtensionsShellEx.h
Comment (pmiossec) [reply native/GitExtensionsShellEx/Generated/GitExtensionsShellEx.h]: No strong feelings: I find this `/src/native` cleaner and more meaningful (i.e. better organized) but maybe a little less convenient.
I think I have a little preference for `/src/native`...

**Comment by RussKie** [comment]
> > Rename Bin/ into assets/ and move under setup/
> 
> 💯 "Bin" was always something to be careful when cleaning manually .net "bin" folders. But I don't like to have a folder "assets" (so with the same name) remaining at the root of the folder and so that could be confusing. Remove this gource image that doesn't bring added value? (and fix `gouurce` to `gource` in `gource video.md`)

That's my plan to move that markdown and the gif to the wiki. These have little use or relevance in the repo itself.

> > Move scripts/ to eng/
> 
> I absolutely don't have a clue of what "eng" could means: so either I will learn a new word/meaning or I'm not very convinced....

"eng" is short for "engineering". That's a convention in the dotnet/* repos and many other MS internal repos. So for me that's kind of natural... Do you have any other preference?

**Comment by RussKie** [comment]
The latest repo layout:
image

**Comment by pmiossec** [comment]
> "eng" is short for "engineering". That's a convention in the dotnet/* repos and many other MS internal repos. So for me that's kind of natural... Do you have any other preference?

TIL. Ok for me (even if it makes me think more about "eng(lish)" than "eng(ineering)". No better idea. And let's not reinvent a convention (which, as a consequence, won't be a convention 🤔 )

>That's my plan to move that markdown and the gif to the wiki. These have little use or relevance in the repo itself.

I will be more than happy if it is done in this PR...

**Review by mstv — COMMENTED** [review]
Looks sensible, have not run
I do not see at the first and second glance why the AppVeyor build timed out > 4 minutes after "build success".

image

---

> Crossing the fingers very hard that Blame is not "broken"

It worked well this time except for a few MSVS project files - if squashed into a single commit.

---

The "sorting" feature was a great help for the review:

image

**Comment by RussKie** [comment]
> The "sorting" feature was a great help for the review:
> 
> image

TIL 🔥

**Comment by RussKie** [comment]
This is ready. Unless there are objections I'm planning to merge it tomorrow.

**Comment by RussKie** [comment]
> > That's my plan to move that markdown and the gif to the wiki. These have little use or relevance in the repo itself.
> 
> I will be more than happy if it is done in this PR...

Done. https://github.com/gitextensions/gitextensions/wiki/How-to-visualize-Git-repository's-history

**Review by mstv — APPROVED** [review]
Git is able to automatically resolve the MC when rebasing

**Review by gerhardol — APPROVED** [review]

**Comment by pmiossec** [comment]
> Git is able to automatically resolve the MC when rebasing

They are some cases where Git is not good:
* new files added in commits rebased. Because Git handling only files and not directory rename, it still try to create the files in the old directory.
* code moved in another file or somewhere else in the same file and modified in the other branch. Most of the times, changes have to be reapplied manualy.

**Review by pmiossec — APPROVED** [review]

**Comment by mstv** [comment]
> They are some cases where Git is not good:

We know - unfortunately.

>> Git is able to automatically resolve the MC when rebasing

**the** MC (of **this** PR)

**Comment by pmiossec** [comment]
I wonder if this PR does not break the `update-loc.cmd`. I have this error now:

```
C:\Users\pmios\Documents\Perso\Dev\GitHub\gitextensions\src\plugins\Statistics\GitStatistics\GitExtensions.Plugins.GitS
tatistics.csproj : error MSB4057: The target "_UpdateEnglishTranslations" does not exist in the project.
C:\Users\pmios\Documents\Perso\Dev\GitHub\gitextensions\src\plugins\Gource\GitExtensions.Plugins.Gource.csproj : error
MSB4057: The target "_UpdateEnglishTranslations" does not exist in the project.
```

**Comment by pmiossec** [comment]
Forget my comment. Found the issue, 1 min later....... 🙄

**Pull request merged**
